server/src/test/java/org/elasticsearch/action/support/SubscribableListenerTests.java

```diff
@@ -604,4 +606,87 @@ public void testAndThenAcceptFailure() {
         assertComplete(chainedListener, "simulated");
     }
+
+    public void testJavaDocExample() {
+        safeAwait(SubscribableListener.<Boolean>newForked(l -> exampleAsyncMethod(randomIdentifier(), List.of(), l)));
+    }
+
+    private void exampleAsyncMethod(String request, List<Long> items, ActionListener<Boolean> finalListener) {
+        SubscribableListener
+
+            // Start the chain and run the first step by creating a SubscribableListener using newForked():
+            .<String>newForked(l -> firstAsyncStep(request, l))
+
+            // Run a second step when the first step completes using andThen(); if the first step fails then the exception falls through to
+            // the end without executing the intervening steps.
+            .<Integer>andThen((l, firstStepResult) -> secondAsyncStep(request, firstStepResult, l))
+
+            // Run another step when the second step completes with another andThen() call; as above this only runs if the first two steps
+            // succeed.
+            .<Boolean>andThen((l, secondStepResult) -> {
+                if (condition) {
+                    // Steps are exception-safe: an exception thrown here will be passed to the listener rather than escaping to the
+                    // caller.
+                    throw new IOException("failure");
+                }
+
+                // Steps can fan out to multiple subsidiary async actions using utilities like RefCountingListener.
+                final var result = new AtomicBoolean();
+                try (var listeners = new RefCountingListener(l.map(v -> result.get()))) {
+                    for (final var item : items) {
+                        thirdAsyncStep(secondStepResult, item, listeners.acquire());
+                    }
+                }
+            })
+
+            // Synchronous (non-forking) steps which do not return a result can be expressed using andThenAccept() with a consumer:
+            .andThenAccept(thirdStepResult -> {
+                if (condition) {
+                    // andThenAccept() is also exception-safe
+                    throw new ElasticsearchException("some other problem");
+                }
+                consumeThirdStepResult(thirdStepResult);
+            })
+
+            // Synchronous (non-forking) steps which do return a result can be expressed using andThenApply() with a function:
+            .andThenApply(voidFromStep4 -> {
+                if (condition) {
+                    // andThenApply() is also exception-safe
+                    throw new IllegalArgumentException("failure");
+                }
+                return computeFifthStepResult();
+            })
+
+            // To complete the chain, add the outer listener which will be completed with the result of the previous step if all steps were
+            // successful, or the exception if any step failed.
+            .addListener(finalListener);
```

Comment:
It doesn't seem like this test is proving that this chain of logic works as expected unless something is asserted/confirmed at the end? Perhaps pass something through each step, maybe an integer that gets added to, multiplied and/or divided at each step such that we can confirm everything ran in the expected order by the end? Or an array of integers, and each step increments a counter and enters the value in a particular array slot for that step, so we can check at the end that the array entries strictly sequentially increase from step to step and there are no unanticipated orderings.

I'm not sure if this is overkill, or reasonable.

Reply:
The `safeAwait` in the test is asserting that the listener is completed successfully, and that's enough for this test IMO.

Reply:
Okay, sounds good 👍 

Ah, I didn't notice what `safeAwait` was doing.